src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -116,18 +120,56 @@ static bool MultiRouterPlannableQuery(Query *query,
 static RelationRestrictionContext * CopyRelationRestrictionContext(
 	RelationRestrictionContext *oldContext);
 static Node * InstantiatePartitionQual(Node *node, void *context);
-static void ErrorIfInsertSelectQueryNotSupported(Query *queryTree,
-												 RangeTblEntry *insertRte,
-												 RangeTblEntry *subqueryRte,
-												 bool allReferenceTables);
-static void ErrorIfMultiTaskRouterSelectQueryUnsupported(Query *query);
-static void ErrorIfInsertPartitionColumnDoesNotMatchSelect(Query *query,
-														   RangeTblEntry *insertRte,
-														   RangeTblEntry *subqueryRte,
-														   Oid *
-														   selectPartitionColumnTableId);
+static DeferredErrorMessage * InsertSelectQuerySupported(Query *queryTree,
+														 RangeTblEntry *insertRte,
+														 RangeTblEntry *subqueryRte,
+														 bool allReferenceTables);
+static DeferredErrorMessage * MultiTaskRouterSelectQuerySupported(Query *query);
+static DeferredErrorMessage * InsertPartitionColumnMatchesSelect(Query *query,
+																 RangeTblEntry *insertRte,
+																 RangeTblEntry *
+																 subqueryRte,
+																 Oid *
+																 selectPartitionColumnTableId);
 static void AddUninstantiatedEqualityQual(Query *query, Var *targetPartitionColumnVar);
-static void ErrorIfQueryHasModifyingCTE(Query *queryTree);
+static DeferredErrorMessage * ErrorIfQueryHasModifyingCTE(Query *queryTree);
+
+
+DeferredErrorMessage *
+CreateRouterPlan(MultiPlan **newPlan,
+				 Query *originalQuery, Query *query,
+				 RelationRestrictionContext *restrictionContext)
+{
+	if (MultiRouterPlannableQuery(query, restrictionContext))
+	{
+		return CreateSingleTaskRouterPlan(newPlan, originalQuery, query,
+										  restrictionContext);
+	}
+
+	/* FIXME: error? description */
+	return NULL;
```

Comment:
Well, that'd not really change the issue that it'd be nice to have an error message to report.  We'll right now only show the !router error message. But there's been clear desire to have a guc that only allows router queries, and it'd also be nice if we'd display both error messages when both router and distributed planner fail.  Since we don't really know what query was intended.